src/compilation/mod.rs

```diff
@@ -31,31 +36,37 @@ impl<'a> JSONSchema<'a> {
     /// Compile the input schema into a validation tree
     pub fn compile(
         schema: &'a Value,
-        draft: Option<schemas::Draft>,
+        config: Option<JSONSchemaCompileConfig>,
     ) -> Result<JSONSchema<'a>, CompilationError> {
         // Draft is detected in the following precedence order:
         //   - Explicitly specified;
         //   - $schema field in the document;
         //   - Draft7;
-        let draft = draft.unwrap_or_else(|| {
-            schemas::draft_from_schema(schema).unwrap_or(schemas::Draft::Draft7)
-        });
+
+        let mut config = config.unwrap_or_default();
+        config.set_draft_if_missing(schema);
+        let processed_config: Cow<'_, JSONSchemaCompileConfig> = Cow::Owned(config);
+        let draft = processed_config.draft();
+
         let scope = match schemas::id_of(draft, schema) {
             Some(url) => url::Url::parse(url)?,
             None => DEFAULT_SCOPE.clone(),
         };
         let resolver = Resolver::new(draft, &scope, schema)?;
-        let context = CompilationContext::new(scope, draft);
+        let context = CompilationContext::new(scope, processed_config);
 
-        let mut validators = compile_validators(schema, &context)?;
-        validators.shrink_to_fit();
-
-        Ok(JSONSchema {
-            draft,
+        let mut jsonschema = JSONSchema {
             schema,
             resolver,
-            validators,
-        })
+            validators: Vec::with_capacity(0),
+            context,
+        };
+
+        let mut validators = compile_validators(schema, &jsonschema.context)?;
```

Comment:
It seems more clear to create `JSONSchema` in the end, rather than modifying it after creation, especially for this case - it is unconditionally replaced

Reply:
Yes, that's correct I was experimenting with something related to lifetimes and I ended up keeping in the repo instead of discarding the change